lib/spack/spack/builder.py

```diff
@@ -0,0 +1,564 @@
+        def _adapter(self):
+            return getattr(self.pkg_with_dispatcher, attribute_name)
+
+        return property(_adapter)
+
+    @staticmethod
+    def combine_callbacks(pipeline_attribute_name):
+        """This function combines callbacks from old-style packages with callbacks that might
+        be registered for the default builder.
+
+        It works by:
+        1. Extracting the callbacks from the old-style package
+        2. Transforming those callbacks by adding an adapter that receives a builder as argument
+           and calls the wrapped function with ``builder.pkg``
+        3. Combining the list of transformed callbacks with those that might be present in the
+           default builder
+        """
+
+        def _adapter(self):
+            def unwrap_pkg(fn):
+                @functools.wraps(fn)
+                def _wrapped(builder):
+                    return fn(builder.pkg_with_dispatcher)
+
+                return _wrapped
+
+            # Concatenate the current list with the one from package
+            callbacks_from_package = getattr(self.pkg, pipeline_attribute_name, [])
+            callbacks_from_package = [(key, unwrap_pkg(x)) for key, x in callbacks_from_package]
+            callbacks_from_builder = getattr(super(type(self), self), pipeline_attribute_name, [])
+            return callbacks_from_package + callbacks_from_builder
+
+        return property(_adapter)
+
+    def __new__(mcs, name, bases, attr_dict):
+        # Add ways to intercept methods and attribute calls and dispatch
+        # them first to a package object
+        default_builder_cls = bases[0]
+        for phase_name in default_builder_cls.phases:
+            attr_dict[phase_name] = _PackageAdapterMeta.phase_method_adapter(phase_name)
+
+        for method_name in default_builder_cls.legacy_methods:
+            attr_dict[method_name] = _PackageAdapterMeta.legacy_method_adapter(method_name)
+
+        for attribute_name in default_builder_cls.legacy_attributes:
+            attr_dict[attribute_name] = _PackageAdapterMeta.legacy_attribute_adapter(
+                attribute_name
+            )
+
+        #
+        combine_callbacks = _PackageAdapterMeta.combine_callbacks
+        attr_dict[_RUN_BEFORE.attribute_name] = combine_callbacks(_RUN_BEFORE.attribute_name)
+        attr_dict[_RUN_AFTER.attribute_name] = combine_callbacks(_RUN_AFTER.attribute_name)
+
+        return super(_PackageAdapterMeta, mcs).__new__(mcs, name, bases, attr_dict)
+
+
+class InstallationPhase(object):
+    """Manages a single phase of the installation.
+
+    This descriptor stores at creation time the name of the method it should
+    search for execution. The method is retrieved at __get__ time, so that
+    it can be overridden by subclasses of whatever class declared the phases.
+
+    It also provides hooks to execute arbitrary callbacks before and after
+    the phase.
+    """
+
+    # TODO: Add pluggable logging to this object
```

Comment:
To my knowledge, logging is already managed by `installer.py`. Could you expand on what is missing here? (I looked around and didn't see this comment moved from anywhere, so I assume you added it).

Reply:
Leftover comment, I'll remove it. I was wondering at the beginning if I might use a constructor like:
```python
def __init__(self, name, builder, logger=None):
```
in case I needed logging within a phase. We're always in time to add it in the future.